src/kernels/composable_kernel/include/kernel_algorithm/gridwise_generic_2d_reduction_multiblock.hpp

```diff
@@ -0,0 +1,305 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef CK_GRIDWISE_GENERIC_2D_REDUCTION_MULTIBLOCK_HPP
+#define CK_GRIDWISE_GENERIC_2D_REDUCTION_MULTIBLOCK_HPP
+
+#include "float_type.hpp"
+#include "reduction_operator.hpp"
+#include "reduction_functions.hpp"
+#include "reduction_common.hpp"
+
+#include "blockwise_generic_tensor_slice_copy.hpp"
+#include "ConstantMatrixDescriptor.hpp"
+
+namespace ck {
+
+template <int BlockSize,
```

Comment:
index_t

Reply:
> @fronteer Can you explain what is the relation between functions in `reduction_functions.hpp` and other reduction files like `gridwise_generic_reduction.hpp`. I'm not sure if it is necessary to create this many layers of function calls

The reduction_functions.hpp is needed to make the codes easier to maintain.   The complication is caused by the need of handling the following situations: 
     1)  No indices recording verse indices recording 
          * Reduction with indices recorded may need more consideration beside recording the indices, because we need to 
             have the recorded indices consistent as the sequential reduction (eg. if value(A[1]) == value(A[8]), we always have A[1] recorded ) 

     2)  First-time reduction verse second-time reduction. 
           * The second-time reduction is called when the first-time reduction using "Multi-block" method. 
           * The second-time reduction need to read the recorded indices from the first-time if indices are recorded

     3)   Without nan-propagation verse with nan-propagation